### PR TITLE
docs: Document running at massive scale. Fixes #6443

### LINKED ISSUE #6443: Document running at scale
We haven't documented this and users have been asking for it.

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by alexec**
Signed-off-by: Alex Collins <alex_collins@intuit.com>
Fixes #6443

### CODE REVIEW

File: docs/running-at-massive-scale.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+## Test Your Cluster Before You Install Argo Workflows
+
+You'll need a big cluster, with a big Kubernetes master.
+
+Users often encounter problems with Kubernetes needing to be configured for the scale. E.g. Kubernetes API server being
+too small. We recommend you test your cluster to make sure it can run the number of pods they need, even before
+installing Argo. Create pods at the rate you expect that it'll be created in production. Make sure Kubernetes can keep
+up with requests to delete pods at the same rate.
+
+You'll need to to GC data quickly. The less data that Kubernetes and Argo deal with, the less work they need to do. Use
```
Comment (terrytangyuan) [inline docs/running-at-massive-scale.md:20]:
```suggestion
You'll need to GC data quickly. The less data that Kubernetes and Argo deal with, the less work they need to do. Use
```

File: docs/running-at-massive-scale.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+pod GC and workflow GC to achieve this.
+
+## Overwhelmed Kubernetes API
+
+Where Argo has a lot of work to do, the Kubernetes API can be overwhelmed. There are several strategies to reduce this:
+
+* Use the Emissary executor (>= v3.1). This does not make any Kubernetes API requests.
+* Limit the number of concurrent workflows using parallelism.
+* Rate-limit pod creation [configuration](workflow-controller-configmap.yaml) (>= v3.1).
+* Set `DEFAULT_REQUEUE_TIME=1m`.
```
Comment (terrytangyuan) [inline docs/running-at-massive-scale.md:30]: Link to https://github.com/argoproj/argo-workflows/blob/master/docs/environment-variables.md for how to set this?

File: docs/running-at-massive-scale.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+up with requests to delete pods at the same rate.
+
+You'll need to to GC data quickly. The less data that Kubernetes and Argo deal with, the less work they need to do. Use
+pod GC and workflow GC to achieve this.
+
+## Overwhelmed Kubernetes API
+
+Where Argo has a lot of work to do, the Kubernetes API can be overwhelmed. There are several strategies to reduce this:
+
+* Use the Emissary executor (>= v3.1). This does not make any Kubernetes API requests.
```
Comment (terrytangyuan) [inline docs/running-at-massive-scale.md:27]: Unless the templates are k8s custom resources

File: docs/running-at-massive-scale.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
+# Running At Massive Scale
+
+Argo Workflows is capable of running 1,000s of workflows a day, each with 10,000s of nodes. But you'll need to do some
```
Comment (terrytangyuan) [inline docs/running-at-massive-scale.md:3]: Link to the survey or stress test suite?

File: docs/running-at-massive-scale.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+* Use the Emissary executor (>= v3.1). This does not make any Kubernetes API requests.
+* Limit the number of concurrent workflows using parallelism.
+* Rate-limit pod creation [configuration](workflow-controller-configmap.yaml) (>= v3.1).
+* Set `DEFAULT_REQUEUE_TIME=1m`.
+
+## Overwhelmed Database
+
+If you're running workflows with many nodes, you'll probably be offloading data to a database. Offloaded data is kept
+for 5m. You can reduce the number of records create by setting `DEFAULT_REQUEUE_TIME=1m`. This will slow reconciliation,
+but will suite workflows where nodes run for over 1m.
```
Comment (mac9416) [inline docs/running-at-massive-scale.md:36]:
```suggestion
but will suit workflows where nodes run for over 1m.
```

Suit, i.e., "will be appropriate for"?

**Pull request merged**